pyccel/ast/numpyext.py

```diff
@@ -890,9 +894,10 @@ def __init__(self, shape, fill_value, dtype=None, order='C'):
         # If there is no dtype, extract it from fill_value
         # TODO: must get dtype from an annotated node
         if not dtype:
-            dtype = fill_value.dtype
-
+            dtype = DtypePrecisionToCastFunction[fill_value.dtype.name][fill_value.precision]
         # Verify dtype and get precision
+        if dtype in NativeNumeric:
+            dtype = DtypePrecisionToCastFunction[dtype.name][-1]
```

Comment:
Why is this needed?

Reply:
In the way i implemented it it's needed because `NumpyFill` uses a `NativeNumeric` as a default value, this could be avoided if i reverted the default value back to an `str`.

Reply:
I think that would be more suited actually.

Reply:
`PythonFloat` is not a `NativeNumeric`

Reply:
Reverting the default value is also fine though. I think the original default is clearer

Reply:
Oh yeah, that one's from the original change i forgot to remove.

Reply:
When i checked that condition again, i remembered that i left it because it solved the issue of classes like `NumpyImag`,
passing a `NativeNumeric` type as a specified dtype.

https://github.com/pyccel/pyccel/blob/a88889df050aed67343c06cc38532612b017ecca/pyccel/ast/numpyext.py#L216-L231

Reply:
Ah ok. That makes sense. Does `NumpyImag` set the correct precision though? Presumably this handling means it also uses default precision even if we are taking the imaginary part of a `complex64`?

We need to be sure if there is an error because we are missing the handling of `NumericNative` or because the function triggering the error isn't being clear enough about the requested precision

Reply:
Fixed in [4d429e8](https://github.com/pyccel/pyccel/pull/1166/commits/4d429e8cee532f46d505f09b3c295b7852c9b27b).